text/camlp5-associativity.md

```diff
@@ -0,0 +1,146 @@
+
+Note that in this case, rules with right associativity (i.e. ending with `SELF LEVEL "foo"`) yield precedence (i.e. bind wider) than rules with left associativity.
+
+See e.g. coq/coq#17859 and camlp5/camlp5#100 for more details.
+
+Note that an example of conflicts on setting different associativities to the same level is told to be when combining Iris and MathComp.
+
+## Camlp5 associativity shown with `Print Grammar constr` leads to confusion
+
+The `RIGHTA`/`LEFTA` status of a level is thus indicative, and not constraining. And, actually, Coq do override it by setting itself the right-hand side of rules to `NEXT` or `SELF LEVEL "foo"`.
+
+When no official associativity is assigned to a level, Camlp5 decides it is left associative while the Coq table decides (confusingly) that it is right associative. Since the associativity is only a default value for rules which can be overriden per rule, one confusingly sees a level reported as left associative while Coq allows to add only right associative rules to the level (see coq/coq#12373).
+
+Also, the current Camlp5 implementation does not treat associativity in the presence of `LIST1` and `LIST0` on the border of a rule. E.g., in `...; LIST1 SELF -> ...`, `SELF` is interpreted as `SELF TOP` and thus does not follow associativity. This is visible e.g. in the `"choice"; LIST1 rewstrategy` rule for `rewstrategy`.
+
+## Camlp5/gramlib does not fully respect levels
+
+When no rule matches, there is a recovery mechanism (see coq/coq#9008) made of 4 fallbacks that bypasses the level specified by the user.
+
+For instance, `True \/ forall x, x = 0` is tolerated even though `forall` is at level 200 and `/\` at level 80. Or `x~0%Z` is tolerated even though `%Z` expects its arguments to be at level 1.
+
+# Proposal
+
+We propose to go towards a new model:
+- that respects levels
+- that respects no associativity
+- is more flexible about mixing different associativities in the same level
+- remains mostly compatible with the existing situation
+
+# Solutions
+
+## Removal of the recovery mechanism
+
+Removing the recovery mechanism breaks quite common situations such as `True \/ forall x, x = 0`, or `2^-n`, or `x~0%Z`.
+
+One could try to automatically adapt the scripts but I believe that the simplest is to move `forall` (and all `binder_constr` rules spanning until the end of the sentence) to a lower level, such as 10, just above the level of application.
+
+The drawback is that we need to modify the printer so that it knows how to print such cycle made of a rule at some level (here 10) which is lower than the level on the right-hand side of the rule (here 200). This is however already made possible by PR #17875.
+
+For `-`, there are several solutions, such as:
+- to set it at level 10 with arguments at level 40 (the level of `*`) so that `2 ^ - x * y` is `2 ^ - (x * y)` but `2 ^ - x + y` is `(2 ^ - x) + y` (??)
```

Comment:
Parsing `2 ^ - x * y` as `2 ^ (- (x * y))` seems to be quite surprising. I would expect the consensus among mathematicians to be `(2 ^ (- x)) * y`.

In other words, there is no reason for the argument of prefix `"-"` to be at level 40. It should be at the current level (i.e., 35 if the argument of infix `"^"`. In fact, the notion of levels for a notation that starts with a terminal and ends with a non-terminal does not make sense in the first place. It should not have an intrinsic level, it should always use the ambient level.

Reply:
I got confused by the idea that `- x * y` was `- (x * y)` by default  in mathematics, but that's the same as `(- x) * y`, so it does not have to be the former. And actually, most programming languages seem to make prefix `-` binding tightly (so morally at 10 for Coq).

For the example I mentioned, the level of `x` in `- x` do not matter. The whole notation could be at 10 with `x` at 10 too.

What do you mean by "It should not have an intrinsic level, it should always use the ambient level.". That is, in the following examples, how would you associate (or fail):
- `x * - y `
- `x ^ - y`
- `x + - y`
- `f x - y`
- `- f * x`
- `- f  ^ x`
- `- f + x`
- `- f x`

Reply:
> And actually, most programming languages seem to make prefix `-` binding tightly (so morally at 10 for Coq).

That does not mean that it is a good idea. In fact, the very first sentence of the Wikipedia page you cite says that `- 3 ^ 2` is interpreted as `-9` in mathematics. And the only "languages" it mentions as binding tightly are Excel, PlanMaker, and bc.

So, having unary minus at level 10 would be awful. And having it at level 35 is already painful. I hate having to write `- (x * y)`. It should have been at level 45.

> * `x * - y`
> * `x ^ - y`
> * `x + - y`

All three of them associate as `x ! (- y)`.

> * `f x - y`

Do you have an infix minus in the context? If not, then parsing can be recovered as `f x (- y)`.

But the interesting case is actually `f x - y z`, which would be parsed as `f x (- y) z` (assuming again that there is no infix minus, otherwise the point is moot).

> * `- f * x`
> * `- f  ^ x`
> * `- f + x`
> * `- f x`

You did not say what the ambient level is, so it depends. Let us assume that those expressions all look like `y + - f ! x` (so, ambient level is 50 left associative). Then they would respectively parse as

* `y + (- (f * x))`
* `y + (- (f ^ x))`
* `(y + (- f)) + x`
* `y + (- (f x))`

Reply:
(I'm tired, the first 3 questions were pointless. I meant the minus sign in the front)

I don't see what conclusion to draw otherwise. And we are somehow constrained by what is currently done, which is:
```coq
Require Import ZArith. Context (x y :Z) (f : Z -> Z) (g : Z). Set Printing Parentheses. Open Scope Z_scope.
Check  x * - f y. (* x * (- (f y)) *) (* due to recovery *)
Check  x ^ - f y. (* x ^ (- (f y)) *) (* due to recovery *)
Check  x + - f y. (* x + (- (f  y) *) (* due to recovery *)
Check  f x - y.   (* (f  x) - y *)
Check  - g * x.   (* (- g) * x *)
Check  - g ^ x.   (* - (g ^  x) *)
Check  - g + x.   (* (- g) ^  x *)
Check  - f x.     (* - (f x) *)
Check  - x * y.   (* (- x) * y *)
Check  - x ^ y.   (* - (x ^ y) *)
Check  - x + y.   (* (- x) + y *)
```

Reply:
For the record, experimenting changing associativity of:
- `- x * y` from `(- x) * y` to `- (x * y)` and
- `- x / y` from `(- x) / y` to `- (x / y)`

requires several changes in stdlib, starting with:
- `Ring_theory.Ropp_mul_l : -(x * y) == -x * y` (also in `Ncring.v`) and `Ring_theory.ARopp_mul_l : -(x * y) == (-x) * y`
- `BinInt.Zopp_mult_distr_l : - (n * m) = - n * m`
- `Zdivfloor.div_opp_opp : b~=0 -> -a/-b == a/b`
- `Field_theory.rdiv5 : - (a / b) == - a / b`
- `Uint63.opp_spec : φ (- x) = - φ x mod wB`
- `-a÷b` in `Zquote.v`
- `- Qnum x * Zpos (Qden y)` in `QArith_base.v`
- non exhaustive list...